File: python/scan_line.py
'''
Author: Yanrui Hu
Date: 2022/9/16
Description: About numpy's usage.
Version: 0.1
Keywords: scan_line, python, "haven't seen"

'''

# Leetcode 850 矩形面积 II
# Keywords: 扫描线

"""
You are given a 2D array of axis-aligned rectangles. Each rectangle[i] = [xi1, yi1, xi2, yi2] denotes the ith rectangle where (xi1, yi1) are the coordinates of the bottom-left corner, and (xi2, yi2) are the coordinates of the top-right corner.

Calculate the total area covered by all rectangles in the plane. Any area covered by two or more rectangles should only be counted once.

Return the total area. Since the answer may be too large, return it modulo 109 + 7.

来源：力扣（LeetCode）
链接：https://leetcode.cn/problems/rectangle-area-ii
著作权归领扣网络所有。商业转载请联系官方授权，非商业转载请注明出处。
"""

class Solution:
    def rectangleArea(self, rectangles: List[List[int]]) -> int:
        pos = []
        for poses in rectangles:
            pos.extend([poses[0], poses[2]])
        pos.sort()
        ans = 0
        for i in range(1, len(pos)):
            a, b = pos[i-1], pos[i]
            width = b - a
            if width == 0:
                continue
            lines = [(poses[1], poses[3]) for poses in rectangles if poses[0] <= a and b <= poses[2]]
            lines.sort() # 这一步我们得到了很多属于矩形的高度区间
            # 下一步是计算这些高度区间的`积分`, 也就是覆盖的长度和
            height, l, r = 0, -1, -1
            for cur in lines:
                if cur[0] > r:
                    height += r-l
                    l, r = cur
                elif cur[1] > r:
                    r = cur[1] # 扩展右边界，这个妙啊

            height += r-l
            ans += height * width

        return ans % int(1e9+7)

File: cpp/scan_line.cpp
// Author: Yanrui Hu
// Date: 2022/9/16
// Description: Add a new question type.
// Version: 0.1
// Keywords: 扫描线, "Review this!"

// From Leetcode 850 矩形面积 II

#include <algorithm>
#include <cstdint>
#include <functional>
#include <memory>
#include <numeric>
#include <utility>
#include <vector>

using std::pair;
using std::tuple;
using std::vector;

class Solution {
 private:
  vector<tuple<int, int>>* _lines;

  void filter(vector<int>& poses, int a, int b) {
    if (poses[0] <= a && b <= poses[2]) {
      _lines->emplace_back(poses[1], poses[3]);
    }
  }

 public:
  int rectangleArea(vector<vector<int>>& rectangles) {
    const int n = rectangles.size();
    vector<int> pos(2 * n);
    int i = 0;
    for (auto& it : rectangles) {
      pos[i++] = it[0];
      pos[i++] = it[2];
    }
    std::sort(pos.begin(), pos.end());

    std::int64_t ans = 0;

    const int n2 = 2 * n;
    for (int i = 1; i < n2; ++i) {
      int a = pos[i - 1];
      int b = pos[i];
      int width = b - a;

      if (width == 0) continue;

      vector<tuple<int, int>> lines;
      _lines = &lines;

      for_each(rectangles.begin(), rectangles.end(), this->filter);

      sort(lines.begin(), lines.end());
    }
  }
};
